src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManager.cs

```diff
@@ -0,0 +1,481 @@
+                cancellationToken);
+        }
+    }
+
+    /// <summary>
+    /// Synchronizes given repo in VMR onto given revision.
+    /// </summary>
+    private async Task UpdateRepoToRevision(
+        SourceMapping mapping,
+        string fromRevision,
+        string toRevision,
+        string clonePath,
+        bool ignoreWorkingTree,
+        string commitMessage,
+        Signature author,
+        CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var patchPath = GetPatchFilePath(mapping);
+        await CreatePatch(mapping, clonePath, fromRevision, toRevision, patchPath);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var info = new FileInfo(patchPath);
+        if (!info.Exists)
+        {
+            throw new InvalidOperationException($"Failed to find the patch at {patchPath}");
+        }
+
+        if (info.Length == 0)
+        {
+            _logger.LogInformation("No changes for {repo} in given commits (maybe only excluded files changed?)", mapping.Name);
+        }
+        else
+        {
+            await ApplyPatch(mapping, patchPath, cleanWorkingTree: !ignoreWorkingTree);
+        }
+
+        await TagRepo(mapping, toRevision);
+
+        Commit(commitMessage, author);
+    }
+
+    /// <summary>
+    /// Prepares a git repository either by cloning it or if clone is already found, pulling the newest changes.
+    /// </summary>
+    /// <param name="mapping">Repository mapping</param>
+    /// <returns>Path to the cloned repo</returns>
+    private async Task<string> PrepareClone(SourceMapping mapping)
+    {
+        var clonePath = Path.Combine(_tmpPath, mapping.Name);
+        if (Directory.Exists(clonePath))
+        {
+            _logger.LogInformation("Clone of {repo} found, pulling new changes...", mapping.DefaultRemote);
+
+            var localRepo = new LocalGitClient(_processManager.GitExecutable, _logger);
+            localRepo.Checkout(clonePath, mapping.DefaultRef);
+
+            var result = await _processManager.ExecuteGit(clonePath, "pull");
+            result.ThrowIfFailed($"Failed to pull new changes from {mapping.DefaultRemote} into {clonePath}");
+            _logger.LogDebug("{output}", result.ToString());
+
+            return Path.Combine(clonePath, ".git");
+        }
+
+        var remoteRepo = await _remoteFactory.GetRemoteAsync(mapping.DefaultRemote, _logger);
+        remoteRepo.Clone(mapping.DefaultRemote, mapping.DefaultRef, clonePath, checkoutSubmodules: false, null);
+
+        return clonePath;
+    }
+
+    /// <summary>
+    /// Notes down the current SHA the given repo is synchronized to into a file inside the VMR.
+    /// </summary>
+    /// <param name="mapping">Repository</param>
+    /// <param name="commitId">SHA</param>
+    private async Task TagRepo(SourceMapping mapping, string commitId)
+    {
+        var tagFile = Path.Combine(_reposPath, $".{mapping.Name}");
```

Comment:
We should put these away, e.g. `src/.tags` so that `src` only has `source-mappings.json` and the repo directories